src/main/java/bio/terra/service/tabulardata/google/BigQueryPdao.java

```diff
@@ -1047,4 +1052,113 @@ private LegacySQLTypeName translateType(String datatype) {
+        "INSERT INTO `<project>.<dataset>.<softDeleteTable>` " +
+        "SELECT DISTINCT <rowId> FROM `<project>.<dataset>.<softDeleteExtTable>`";
+
+    /**
+     * Insert row ids into the corresponding soft delete table for each table provided.
+     *
+     * @param dataset repo dataset that we are deleting data from
+     * @param tables list of tables that should have corresponding external tables with row ids to soft delete
+     * @param suffix a bq-safe version of the flight id to prevent different flights from stepping on each other
+     */
+    public TableResult applySoftDeletes(Dataset dataset,
+                                 List<DataDeletionTableModel> tables,
+                                 String suffix) {
+        BigQueryProject bigQueryProject = bigQueryProjectForDataset(dataset);
+
+        // we want this soft delete operation to be atomic, we we will combine all of the inserts into one statement
+        // that we send to bigquery.
+        List<String> tableNames = tables
+            .stream()
+            .map(DataDeletionTableModel::getTableName)
+            .collect(Collectors.toList());
+
+        // the soft delete tables have a random suffix on them, we need to fetch those from the db and pass them in
+        Map<String, String> softDeleteTableNameLookup = dataset.getTables()
+            .stream()
+            .collect(Collectors.toMap(DatasetTable::getName, DatasetTable::getSoftDeleteTableName));
+
+        List<String> sqlStatements = tableNames.stream()
+            .map(tableName -> new ST(insertSoftDeleteTemplate)
+                .add("project", bigQueryProject.getProjectId())
+                .add("dataset", prefixName(dataset.getName()))
+                .add("softDeleteTable", softDeleteTableNameLookup.get(tableName))
+                .add("rowId", PDAO_ROW_ID_COLUMN)
+                .add("softDeleteExtTable", externalTableName(tableName, suffix))
+                .render())
+            .collect(Collectors.toList());
+
+        return bigQueryProject.query(String.join(";", sqlStatements));
+    }
+
+    private long getSingleLongValue(TableResult result) {
+        FieldValueList fieldValues = result.getValues().iterator().next();
+        return fieldValues.get(0).getLongValue();
+    }
+
+    /**
+     * This join should pair up every rowId in the external table with a corresponding match in the live view. If there
+     * isn't a match in the live view, then V.rowId will be null and we count that as a mismatch.
+     *
+     * Note that since this is joining against the live view, an attempt to soft delete a rowId that has already been
+     * soft deleted will result in a mismatch.
+     */
+    private static final String validateSoftDeleteTemplate =
+        "SELECT COUNT(E.<rowId>) FROM `<project>.<dataset>.<softDeleteExtTable>` E " +
+        "LEFT JOIN `<project>.<dataset>.<liveView>` V USING (<rowId>) " +
+        "WHERE V.<rowId> IS NULL";
```

Comment:
Joining against the live view instead of the underlying table means that soft deleting the same rowid twice will throw an error, correct? Is that your intention?

Reply:
Yep :)